src/Processors/QueryPlan/BufferChunksTransform.h

```diff
@@ -0,0 +1,30 @@
+#pragma once
+#include <Processors/IProcessor.h>
+#include <queue>
+
+namespace DB
+{
+
+class BufferChunksTransform : public IProcessor
+{
+public:
+    BufferChunksTransform(const Block & header_, size_t max_bytes_to_buffer_, size_t limit_);
```

Comment:
it seems to me that if this optimisation will only be based on the number of bytes we will miss a lot of cases when rows are more or less wide. I think we always can buffer up to a full block for example. and that would be a good lower limit.
wdyt?

Reply:
Yes, probably it makes sense. However in current [implementation](https://github.com/ClickHouse/ClickHouse/blob/ada56bdeeb46a903dbb4756bd61cc213b2d4b5dd/src/Processors/QueryPlan/BufferChunksTransform.cpp#L48) we always buffer at least one chunk because we can exceed the threshold.

Reply:
I meant 65K rows regardless of their size in bytes, because afaiu we assume individual chunks to be much smaller in terms of number of rows